Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -624,12 +630,45 @@ void CGPathApply(CGPathRef path, void* info, CGPathApplierFunction function) {
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 bool CGPathContainsPoint(CGPathRef path, const CGAffineTransform* m, CGPoint point, bool eoFill) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    if (m) {
+        point = CGPointApplyAffineTransform(point, *m);
+    }
+    // check if the point is outside this box already, if it is, return false
+    CGRect boundingBox = CGPathGetBoundingBox(path);
+
+    if ((boundingBox.origin.x > point.x) || (boundingBox.origin.y > point.y) ||
+        ((boundingBox.origin.x + boundingBox.size.width) < point.x) || ((boundingBox.origin.y + boundingBox.size.height) < point.y)) {
+        return false;
+    }
+
+    CGContextRef context =
+        CGBitmapContextCreate(0, boundingBox.origin.x + boundingBox.size.width, boundingBox.origin.y + boundingBox.size.height, 1, 1, 0, 0);
+
+    CGContextSetGrayFillColor(context, 1.0, 1.0);
+    CGContextAddPath(context, path);
+
+    if (eoFill) {
+        CGContextEOFillPath(context);
+    } else {
+        CGContextFillPath(context);
+    }
+
+    CGImageRef image = context->Backing()->DestImage();
+
+    float r = 0.0f;
+    float g = 0.0f;
+    float b = 0.0f;
+    float a = 0.0f;
+    image->Backing()->GetPixel(point.x, point.y, r, g, b, a);
```

Comment:
This raises an interesting question - it seems to me implementations should try to keep to the Obj-C(++) interface that comes from Apple’s APIs as much as possible. If the Cairo implementation ever changed or was updated, this would mean those implementation changes would only have to happen in the original implementation. However - a case could also be said that this could be more efficient (though compiler optimizations may make that a moot point). What are you feelings about that?
Josh

On Jun 28, 2016, at 1:07 PM, Ben Viglietta <notifications@github.com<mailto:notifications@github.com>> wrote:


In Frameworks/CoreGraphics/CGPath.mm<https://github.com/Microsoft/WinObjC/pull/587#discussion_r68832635>:

> +    CGContextSetGrayFillColor(context, 1.0, 1.0);
> +    CGContextAddPath(context, path);
> +
> +    if (eoFill) {
> +        CGContextEOFillPath(context);
> +    } else {
> +        CGContextFillPath(context);
> +    }
> +
> +    CGImageRef image = context->Backing()->DestImage();
> +
> +    float r = 0.0f;
> +    float g = 0.0f;
> +    float b = 0.0f;
> +    float a = 0.0f;
> +    image->Backing()->GetPixel(point.x, point.y, r, g, b, a);


The Cairo library that CoreGraphics is built on already has a function to do this test: cairo_in_fill. It might be more natural, rather than doing the fill and the color test, to add something like _CGContextIsPointInPath, which itself calls through to cairo_in_fill.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Microsoft/WinObjC/pull/587/files/83c35ea7f13358471f0a3f5483315396816db73e#r68832635>, or mute the thread<https://github.com/notifications/unsubscribe/AJC9AsT5OSyrwXMgOmTFrgnnSIceGRNsks5qQX8NgaJpZM4I_fp1>.